rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/listener/TrainingEvent.java

```diff
@@ -0,0 +1,7 @@
+package org.deeplearning4j.rl4j.learning.listener;
+
+/**
+ * The base definition of all training events
+ */
+public class TrainingEvent {
```

Comment:
In a previous commit, I used generics. I used a different generic for each event (training start, end, epoch start, end) because one might need a different event for training start than for epoch start. Overall, it seemed a bit overkill to me but, on the other hand, classes that will use the listener will look cleaner. I will add it.

Reply:
It should define some sort of generic interface or contain data common to all events then that would make sense.

Reply:
I didn't mean literally a Java `interface`. I mean, the user will get a `TrainingEvent` or `ITrainingEvent`, whatever, doesn't matter, but that has no methods, no data. What are we supposed to do with it?

Reply:
It sounds like users should be able to implement all they need as part of the listeners. Do you have a concrete example where that wouldn't be possible?

Reply:
I couldn't come up with an example where it wouldn't be possible without the params so, I changed the signature to parameter-less methods for `onTrainingStart()` and `onTrainingEnd()`.

Reply:
Sounds good, although it would make sense to pass a few things that listeners might be interested in, like possibly the Learning object that is calling it, for example, see:
https://github.com/eclipse/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/api/BaseTrainingListener.java

Reply:
I see what you mean but in the sync case, there is only `SyncLearning` and in the async case, there are `AsyncLearning`, `AsyncThread` and maybe `AsyncGlobal`. 
What do you think if we design this listener and any future events/listeners in RL4J with this design rule:

> Everything that change (i.e. is a different value or instance) between calls are received through a getter of the event parameter. Everything else should be passed as a ctor argument of the particular listener.

What I mean is this:
```
public class MyCustomTrainingListener implements TrainingListener {
    private final MyCustomLearning learning;
	
    public MyCustomTrainingListener(MyCustomLearning learning) { // Type safety at compile time
        this.learning = learning; 
    }
	
    @Override
    public ListenerResponse onNewEpoch(IEpochTrainingEvent event) {
        DoSomethingWith(learning.MyCustomField);
        return ListenerResponse.CONTINUE;
    }
}
```
instead of having to cast a learning member of event:
```
public class MyCustomTrainingListener implements TrainingListener {

    @Override
    public ListenerResponse onNewEpoch(IEpochTrainingEvent event) {
        MyCustomLearning learning = ((MyCustomEpochTrainingEvent)event).getLearning(); // 'Type safety' at run time
        DoSomethingWith(learning.MyCustomField);
        return ListenerResponse.CONTINUE;
    }
}
```

Also, in the case of onNewEpoch and onEpochTrainingResult, there is no way to find the calling AsyncThread instance, we don't have the thread num or the object. Maybe we should add that.

Reply:
Ah, I see, there is a missing base class for Learning. If we had that, would we still need EpochTrainingEvent? Or is the idea with async learning to keep running the threads in the background even while calling the listeners and so we want to take snapshots of the state?

Reply:
The base of the problem is that we want the same listener be able to fetch the info it needs no matter who calls it and sync and async behaves very differently. We could create an interface that contains what is common between SyncLearning and AsyncThread and pass that to the event but, if a user tries to implement a custom learning algorithm and he needs to pass some info to a custom listener, he should be able to do so without having to add a new member to the caller's state (for example, because it wouldn't make sense from the use case's point of view).
Can you tell me for what reason you would like the EpochTrainingEvent gone?

Reply:
It doesn't need to be gone, but it does feel strange... I think we could create a parent ILearning interface or something with these properties instead and pass that to the listeners? Then they could cast that reference however they need to get more information, when needed, without having to create temporary objects like EpochTrainingEvent.

Reply:
I have removed the 'Event' classes. See comment below for details